src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexMatcher.cs

```diff
@@ -493,65 +501,85 @@ private void DoCheckTimeout(int timeoutOccursAt)
         /// <param name="perThreadData">Per thread data reused between calls.</param>
         public SymbolicMatch FindMatch(bool isMatch, ReadOnlySpan<char> input, int startat, PerThreadData perThreadData)
         {
+            // If we need to perform timeout checks, store the absolute timeout value.
             int timeoutOccursAt = 0;
             if (_checkTimeout)
             {
                 // Using Environment.TickCount for efficiency instead of Stopwatch -- as in the non-DFA case.
                 timeoutOccursAt = Environment.TickCount + (int)(_timeout + 0.5);
             }
 
+            // If we're starting at the end of the input, we don't need to do any work other than
+            // determine whether an empty match is valid, i.e. whether the pattern is "nullable"
+            // given the kinds of characters at and just before the end.
             if (startat == input.Length)
             {
-                // Covers the special-case of an empty match at the end of the input.
                 uint prevKind = GetCharKind(input, startat - 1);
                 uint nextKind = GetCharKind(input, startat);
-
-                bool emptyMatchExists = _pattern.IsNullableFor(CharKind.Context(prevKind, nextKind));
-                return emptyMatchExists ?
+                return _pattern.IsNullableFor(CharKind.Context(prevKind, nextKind)) ?
                     new SymbolicMatch(startat, 0) :
                     SymbolicMatch.NoMatch;
             }
 
-            // Find the first accepting state. Initial start position in the input is i == 0.
-            int i = startat;
-
-            // May return -1 as a legitimate value when the initial state is nullable and startat == 0.
-            // Returns NoMatchExists when there is no match.
-            i = FindFinalStatePosition(input, i, timeoutOccursAt, out int i_q0_A1, out int watchdog);
+            // Phase 1:
+            // Determine whether there is a match by finding the first final state position.  This only tells
+            // us whether there is a match but needn't give us the longest possible match. This may return -1 as
+            // a legitimate value when the initial state is nullable and startat == 0. It returns NoMatchExists (-2)
+            // when there is no match.
+            int i = FindFinalStatePosition(input, startat, timeoutOccursAt, out int matchStartLowBoundary, out int watchdog);
 
+            // If there wasn't a match, we're done.
             if (i == NoMatchExists)
             {
                 return SymbolicMatch.NoMatch;
             }
 
+            // A match exists. If we don't need further details, because IsMatch was used (and thus we don't
```

Comment:
Not related to your PR, but should we probably rename isMatch -> `quick` in order to match RegexRunner fields?

Reply:
Sure.  We could also rename the quick field to be isMatch.  Up to you which direction you want to go.

Reply:
I don’t like isMatch since quick can be used in other cases and not just isMatch, so I will later either rename this to quick or to something else more descriptive.